### DISCUSSION TITLE
Yaw Starting at Wrong Value

### DISCUSSION THREAD

**Discussion by rbhandary** [bitcraze/discussions#396] [Q&A]
Hi! I am currently using the crazyflie 2.1 and am flying a very simple path (fly up to a height of 1 m and then back down to land at its original point). I have been having a persistent problem however where the drone is reporting its yaw as around -17 degrees from the very beginning of the flight which then causes it to rotate and land at a yaw which it thinks is 0 however is now at an angle. I had experienced this problem before but it usually corrected itself after a couple of flights (not sure why but didn't question it since it started working eventually). Just wondering what the problem is and how I can fix it? Let me know if you need more info!

**Comment by krichardsson** [comment]
Could you tell us a bit more about your setup? Do you have any positioning system? Flow deck?

Yaw is a surprisingly complex area :-)

**Reply by rbhandary** [reply]
Thanks for your quick response! I am using 2 V1 lighthouse towers with the lighthouse deck.

**Reply by krichardsson** [reply]
I would expect the yaw to be pretty good in a lighthouse system. The yaw is based on the coordinate system that you used when estimating the geometry of the system. 0 degrees of yaw should be when the Crazyflie is aligned with the x-axis.
Is this not the case?

You can easily verify the orientation of the x-axis by opening the lighthouse tab in the client and move the crazyflie around on the floor until you are on some point where x=0 and y>0.

**Reply by rbhandary** [reply]
The behaviour is pretty strange... This morning after a few flights, the yaw seemed to correct itself and only had an error of about 1 degree. However when I flew it this afternoon, the same problem came up again. When I put the drone at the origin of my co-ordinate system, the lighthouse tab says it is at (0,0,0) and when I put it at (1,0,0) of my physical coordinate system, the lighthouse agrees with these coordinates. The yaw measurement when I put the drone at (0,0,0) is now 70 degrees and when I put it at (1,0,0) it has a yaw of -17 degrees. The drone is facing towards the direction of the x axis in both cases. Not really sure why this is happening and why the yaw is sometimes fine and sometimes isn't?

**Reply by rbhandary** [reply]
Just an update, the problem sometimes resolves itself with a bit of flying (and crashing!) but sometimes persists for a while. When the drone flies directly up and directly down the crazyflie turns (to where it thinks yaw = 0) is and sways a lot before it lands. When the crazyflie is programmed to fly directly up to (0,0,1) then towards the x axis to (1,0,1) and then to land at (1,0,0), it flies up and turns to where it thinks yaw = 0 is and then starts to circle on its flight towards the x axis before it crashes.

**Reply by krichardsson** [reply]
I'm surprised that the yaw is off, but there is always the possibility of a bug :-)
The mechanism that pushes yaw into the estimator is a bit of a hack and could be improved, on the other hand it has been used for years (both for LH1 and 2) and I have not heard about this type of problem before. 

Where are your base stations placed? Is it pretty standard at 2-3 meters height facing towards the center of the flying space, or are they at some more unusual angle?

> Just an update, the problem sometimes resolves itself with a bit of flying (and crashing!) but sometimes persists for a while. When the drone flies directly up and directly down the crazyflie turns (to where it thinks yaw = 0) is and sways a lot before it lands. When the crazyflie is programmed to fly directly up to (0,0,1) then towards the x axis to (1,0,1) and then to land at (1,0,0), it flies up and turns to where it thinks yaw = 0 is and then starts to circle on its flight towards the x axis before it crashes.

What you describe is more expected behavior. What I think is happening is that when you tell the CF to go to some coordinate, the yaw will be used to figure out which direction to go, in your case the CF "thinks" it is facing the direction it should go in and starts to go forward. The estimator will realize that the position is changing in the "wrong" way and start to correct the yaw. 

If you hold the CF in your hand and move it around in your flight space, examining the yaw in the client, can you see some pattern?

You can try to cover one base station at a time and check if you get the same behavior.

Do you at any point see the yaw slowly drift back to 0?

In the console log in the client, do you get a warning about a broken lighthouse sensor? 

Have you tried to re-estimate the geometry in the client?
